src/lib.rs

```diff
@@ -1,17 +1,55 @@
-#![cfg(any(target_os = "macos", target_os = "ios"))]
-#![allow(clippy::missing_safety_doc, clippy::let_unit_value)]
+#![cfg(target_vendor = "apple")]
+#![allow(clippy::missing_safety_doc)]
+#![cfg_attr(docsrs, feature(doc_auto_cfg, doc_cfg_hide), doc(cfg_hide(doc)))]
+#![deny(unsafe_op_in_unsafe_fn)]
 
-#[macro_use]
-extern crate objc;
-
-use objc::runtime::Object;
+use objc2::rc::Retained;
+use objc2_quartz_core::CAMetalLayer;
+use std::ffi::c_void;
 
+#[cfg(any(target_os = "macos", doc))]
 pub mod appkit;
+
+#[cfg(any(not(target_os = "macos"), doc))]
 pub mod uikit;
 
-pub type CAMetalLayer = *mut Object;
+/// A wrapper around [`CAMetalLayer`].
+pub struct Layer {
+    layer: Retained<CAMetalLayer>,
+    pre_existing: bool,
+}
+
+impl Layer {
+    /// Get a pointer to the underlying [`CAMetalLayer`]. The pointer is valid
+    /// for at least as long as the [`Layer`] is valid, but can be extended by
+    /// retaining it.
+    ///
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// use objc2::rc::Retained;
+    /// use objc2_quartz_core::CAMetalLayer;
+    /// use raw_window_metal::Layer;
+    ///
+    /// let layer: Layer;
+    /// # layer = unimplemented!();
+    ///
+    /// let layer: *mut CAMetalLayer = layer.as_ptr().cast();
+    /// // SAFETY: The pointer is a valid `CAMetalLayer`.
+    /// let layer = unsafe { Retained::retain(layer).unwrap() };
+    ///
+    /// // Use the `CAMetalLayer` here.
+    /// ```
+    #[inline]
+    pub fn as_ptr(&self) -> *mut c_void {
+        let ptr: *const CAMetalLayer = Retained::as_ptr(&self.layer);
+        ptr as *mut _
+    }
 
-pub enum Layer {
-    Existing(CAMetalLayer),
-    Allocated(CAMetalLayer),
+    /// Whether `raw-window-metal` created a new [`CAMetalLayer`] for you.
+    #[inline]
+    pub fn pre_existing(&self) -> bool {
+        self.pre_existing
+    }
```

Comment:
I think it can still be useful for the user to know what `raw-window-metal` ended up doing (e.g. I'm considering porting `wgpu` to use this, and there it would be useful to know if we can override the layer's delegate, or if we should refrain from doing so).

Reply:
Good point, see https://github.com/rust-windowing/raw-window-metal/pull/20